## Applied Data Science Coursera Capstone
# Segmenting and Clustering Neighboods in Toronto

For this project, we want to explore and cluster the neighborhoods in Toronto. This exploration will take place in three parts.

* [Part 1: Scrape the Web](#part1)
* [Part 2: Get More Geographic Information](#part2)
* [Part 3: Explore & Cluster the Neighborhods](#part3)
  * [Subset the data based on borough name](#slice_borough)
  * [Using the Foursquare API](#foursquare_api)
  * [Mapping our neighboords with Folium](#folium_boroughs)
  * [Neighborhood Analysis](#neighborhood_analysis)
  * [Building a dataframe of top venues in each neighborhood](#top_5_venues_dataframe)
  * [Clustering the neighborhoods](#clustering_neighborhoods)
  * [Examining the clusters](#examine_clusters)

---

<a id='part1'></a>

## Part 1

We will use a webscraping package to retrieve neighborhood information for Toronto available via a Wikipedia page. We will save the scraped data in a dataframe and clean its content.

We can begin by importing the libraries we will use for this analysis.

In [1]:
# Import libraries.
import pandas as pd                           # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
from bs4 import BeautifulSoup                 # for website scraping
from urllib.request import urlopen            # for working with url requests
import requests                               # library to handle requests
import json                                   # library to handle JSON files
from pandas.io.json import json_normalize     # tranform JSON file into a pandas dataframe
import numpy as np                            # library to handle data in a vectorized manner

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


Since we are using *BeautifulSoup* for our web scraping, we will need to assign our URL for scraping, and build the constructor to work with *BeautifulSoup* objects. This object will be the webpage we want to scrape with its associated content and HTML tags.

In [3]:
# URL for scraping.
webpage = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"

# Build BeautifulSoup constructor.
with urlopen(webpage) as fp:
    soup = BeautifulSoup(fp)

Before we start working on extracting content, let's first initialize the dataframe we want to build, with columns for the postal code, borough, and neighborhood.

In [4]:
# Initialize our dataframe.
df = pd.DataFrame(columns=['PostalCode', 'Borough', 'Neighborhood'])
df

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

So far, so good. Now, we can start getting our hands dirty with extraction.

As we look at the webpage in our browser, right-click, and select *inspect element* from our browswer functions, we can see that the table containing information we want on this page is enclosed within a table in the body of the HTML document.

We can also see that the header is enclosed in ```<th></th>``` tags. Since we have already built our dataframe with this columns, we can ignore these. The data we want within the table is actually enclosed in ```<td></td>``` tags. We can use the ```find_all``` function to search for and retrieve the ```td``` tags.

In [5]:
# Find all entries of table contained within body > table > tbody elements.
web_table = soup.body.table.tbody.find_all('td')

We can see that this method works, but the information we need is contained within a pair of tags. We will need to loop through this object and pull out the text we need using *BeautifulSoup*'s ```contents``` method.

In [6]:
postal_code = []
borough = []
neighborhood = []

for i in range(len(web_table)):
    entry = web_table[i].contents     # the contents between each <td> tag
    entry = entry[0].strip()          # remove any newline characters
    if i % 3 == 0:                    # corresponds to leftmost column index
        postal_code.append(entry)
    if i % 3 == 1:                    # corresponds to middle column index
        borough.append(entry)
    if i % 3 == 2:                    # corresponds to rightmost column index
        neighborhood.append(entry)

Now that we have our lists, we can add them to the dataframe, ```df```, that we initialized before, and view the first ten rows.

In [7]:
df['PostalCode'] = postal_code
df['Borough'] = borough
df['Neighborhood'] = neighborhood
df.head(10)

PostalCode           Borough                                  Neighborhood
0        M1A      Not assigned                                              
1        M2A      Not assigned                                              
2        M3A        North York                                     Parkwoods
3        M4A        North York                              Victoria Village
4        M5A  Downtown Toronto                    Regent Park / Harbourfront
5        M6A        North York             Lawrence Manor / Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
7        M8A      Not assigned                                              
8        M9A         Etobicoke                              Islington Avenue
9        M1B       Scarborough                               Malvern / Rouge

Let's check the shape of our dataframe to see the number of rows and columns it has.

In [8]:
df.shape

(180, 3)

As we can see from the first ten rows, we have some boroughs unassigned and empty neighborhood values. Since we will eventually need borough values to be present, we will begin by filtering out the rows that have values as *Not assigned* for their boroughs. We will create a new dataframe with this filter in place in case we need to refer back to the original data.

In [9]:
df_filtered = df[df['Borough'] != 'Not assigned']
df_filtered.head(10)

PostalCode           Borough                                  Neighborhood
2         M3A        North York                                     Parkwoods
3         M4A        North York                              Victoria Village
4         M5A  Downtown Toronto                    Regent Park / Harbourfront
5         M6A        North York             Lawrence Manor / Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park / Ontario Provincial Government
8         M9A         Etobicoke                              Islington Avenue
9         M1B       Scarborough                               Malvern / Rouge
11        M3B        North York                                     Don Mills
12        M4B         East York              Parkview Hill / Woodbine Gardens
13        M5B  Downtown Toronto                      Garden District, Ryerson

Let's check whether we have any ```Neighborhood``` values that have not been assigned.

In [10]:
df_filtered[df_filtered['Neighborhood'] == '']

Empty DataFrame
Columns: [PostalCode, Borough, Neighborhood]
Index: []

It looks like we do not have any neighborhoods that have not been assigned values.

Now we can work on reformatting the neighborhoods where there are multiple neighborhoods per postal code. We want to replace the slashes with commas.

In [11]:
df_filtered[['Neighborhood']] = df_filtered[['Neighborhood']].replace(" / ", ", ", regex = True)
df_filtered.head(10)

/Users/christopher/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:3494: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  self[k1] = value[k2]


PostalCode           Borough                                 Neighborhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
8         M9A         Etobicoke                             Islington Avenue
9         M1B       Scarborough                               Malvern, Rouge
11        M3B        North York                                    Don Mills
12        M4B         East York              Parkview Hill, Woodbine Gardens
13        M5B  Downtown Toronto                     Garden District, Ryerson

This looks nearly perfect; there appears to be something wrong with our row indexing. Let's fix that.

In [12]:
df_filtered.reset_index(inplace = True)
df_filtered.head(10)

index PostalCode           Borough  \
0      2        M3A        North York   
1      3        M4A        North York   
2      4        M5A  Downtown Toronto   
3      5        M6A        North York   
4      6        M7A  Downtown Toronto   
5      8        M9A         Etobicoke   
6      9        M1B       Scarborough   
7     11        M3B        North York   
8     12        M4B         East York   
9     13        M5B  Downtown Toronto   

                                  Neighborhood  
0                                    Parkwoods  
1                             Victoria Village  
2                    Regent Park, Harbourfront  
3             Lawrence Manor, Lawrence Heights  
4  Queen's Park, Ontario Provincial Government  
5                             Islington Avenue  
6                               Malvern, Rouge  
7                                    Don Mills  
8              Parkview Hill, Woodbine Gardens  
9                     Garden District, Ryerson

It looks like we have our row numbering in the proper order. Now, let's drop the added ```index``` column.

In [13]:
df_filtered = df_filtered.drop('index', axis = 1)
df_filtered.head(10)

PostalCode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government
5        M9A         Etobicoke                             Islington Avenue
6        M1B       Scarborough                               Malvern, Rouge
7        M3B        North York                                    Don Mills
8        M4B         East York              Parkview Hill, Woodbine Gardens
9        M5B  Downtown Toronto                     Garden District, Ryerson

It looks like we're all set to move onto the next section.

Before we do, let's check the shape of our resulting dataframe to see how many rows we ended with after modifications.

In [14]:
df.shape

(180, 3)

---

<a id='part2'></a>

## Part 2

Now that we have our postal codes, borough names, and neighborhood names, we need to get geographical data in the form of latitude and longitude coordinates for each of them.

We can do so by getting the geocodes directly from a .csv file.

In [15]:
geo_file = "https://cocl.us/Geospatial_data/Geospatial_Coordinates.csv"

geo_coords = pd.read_csv(geo_file, header = 0)

After retrieving the file and importing it into a *pandas* dataframe, let's take a look at the first five rows.

In [16]:
geo_coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

We should modify the dataframe so that the ```Postal Code``` column is the index.

In [17]:
geo_coords.set_index('Postal Code', inplace = True)
geo_coords.head()

Latitude  Longitude
Postal Code                      
M1B          43.806686 -79.194353
M1C          43.784535 -79.160497
M1E          43.763573 -79.188711
M1G          43.770992 -79.216917
M1H          43.773136 -79.239476

Now that we have the dataframe with geocodes corresponding to Toronto postal codes, we want to add these to our ```df_filtered``` dataframe.

We begin by setting ```PostalCode``` as our index.

In [18]:
df_filtered.set_index('PostalCode', inplace = True)

In [19]:
geo_lat = []          # save our latitude values
geo_long = []         # save our longitude values

for code in df_filtered.index:          # iterate through the postal codes of our dataframe
    lat = geo_coords.loc[code][0]
    long = geo_coords.loc[code][1]
    geo_lat.append(lat)
    geo_long.append(long)

In [20]:
df_filtered['Latitude'] = geo_lat       # add our latitude values to our dataframe
df_filtered['Longitude'] = geo_long     # add our longitude values to our dataframe
df_filtered.head(10)

Borough                                 Neighborhood  \
PostalCode                                                                  
M3A               North York                                    Parkwoods   
M4A               North York                             Victoria Village   
M5A         Downtown Toronto                    Regent Park, Harbourfront   
M6A               North York             Lawrence Manor, Lawrence Heights   
M7A         Downtown Toronto  Queen's Park, Ontario Provincial Government   
M9A                Etobicoke                             Islington Avenue   
M1B              Scarborough                               Malvern, Rouge   
M3B               North York                                    Don Mills   
M4B                East York              Parkview Hill, Woodbine Gardens   
M5B         Downtown Toronto                     Garden District, Ryerson   

             Latitude  Longitude  
PostalCode                        
M3A         43.753259 -79.329656  
M4A         43.725882 -79.315572  
M5A         43.654260 -79.360636  
M6A         43.718518 -79.464763  
M7A         43.662301 -79.389494  
M9A         43.667856 -79.532242  
M1B         43.806686 -79.194353  
M3B         43.745906 -79.352188  
M4B         43.706397 -79.309937  
M5B         43.657162 -79.378937

Looks good! Once again, we reset our index to replicate our original dataframe style.

In [21]:
df_filtered.reset_index(inplace = True)
df_filtered.head(10)

PostalCode           Borough                                 Neighborhood  \
0        M3A        North York                                    Parkwoods   
1        M4A        North York                             Victoria Village   
2        M5A  Downtown Toronto                    Regent Park, Harbourfront   
3        M6A        North York             Lawrence Manor, Lawrence Heights   
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5        M9A         Etobicoke                             Islington Avenue   
6        M1B       Scarborough                               Malvern, Rouge   
7        M3B        North York                                    Don Mills   
8        M4B         East York              Parkview Hill, Woodbine Gardens   
9        M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494  
5  43.667856 -79.532242  
6  43.806686 -79.194353  
7  43.745906 -79.352188  
8  43.706397 -79.309937  
9  43.657162 -79.378937

Success! We now have a dataframe containing the postal codes, borough names, neighborhood names, and latitude and longitude coordinates of neighborhoods in Toronto.

---
<a id='part3'></a>

## Part 3

Up until now, we have worked towards extracting and shaping the data we needed for exploration. In this section, we can begin the exploration and clustering of certain neighborhoods in Toronto.

Let's start by loading the libraries we will need.

In [22]:
import matplotlib.pyplot as plt    # plotting library
# backend for rendering plots within the browser
%matplotlib inline
from sklearn.cluster import KMeans  # use for clustering
from geopy.geocoders import Nominatim  # convert an address into latitude and longitude values
import folium

<a id="slice_borough"></a>

Rather than focus on all boroughs and neighborhoods in Toronto, we should limit our exploration to certain boroughs. For the remainder of this project, we will focus on boroughs which have the word 'Toronto' in them.

In [23]:
# Filter data on names of boroughs with 'Toronto' in them.
borough_data = df_filtered[df_filtered['Borough'].str.contains('Toronto')]
borough_data.head()

PostalCode           Borough                                 Neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  
2   43.654260 -79.360636  
4   43.662301 -79.389494  
9   43.657162 -79.378937  
15  43.651494 -79.375418  
19  43.676357 -79.293031

In [24]:
borough_data.shape

(39, 5)

In [25]:
# Show the names of the selected boroughs.
borough_names = []

for borough in borough_data['Borough'].unique():
    borough_names.append(borough)
    print(borough)

Downtown Toronto
East Toronto
West Toronto
Central Toronto


Based on our name restriction, there are 39 instances where the word 'Toronto' appears in the boroughs of Toronto. We can also see that these instances consist of four boroughs –– namely, Downtown Toronto, East Toronto, West Toronto, and Central Toronto.

Now, we have two ways we can get geographical coordinates of our boroughs. We can use _Nominatim_ to get geographical coordinates of Toronto, or we can use the latitude and longitude values we imported from our file.

For the time being, we will stick with our first option –– using _Nominatim_.

In [26]:
df_filtered.loc[df_filtered['Borough'] == borough_names[0], 'PostalCode']

2     M5A
4     M7A
9     M5B
15    M5C
20    M5E
24    M5G
25    M6G
30    M5H
36    M5J
42    M5K
48    M5L
80    M5S
84    M5T
87    M5V
91    M4W
92    M5W
96    M4X
97    M5X
99    M4Y
Name: PostalCode, dtype: object

In [27]:
borough_data.loc[borough_data['Borough'] == borough_names[0]].index

Int64Index([2, 4, 9, 15, 20, 24, 25, 30, 36, 42, 48, 80, 84, 87, 91, 92, 96,
            97, 99],
           dtype='int64')

In [28]:
borough_data.index

Int64Index([  2,   4,   9,  15,  19,  20,  24,  25,  30,  31,  36,  37,  41,
             42,  43,  47,  48,  54,  61,  62,  67,  68,  69,  73,  74,  75,
             79,  80,  81,  83,  84,  86,  87,  91,  92,  96,  97,  99, 100],
           dtype='int64')

In [29]:
city_address = ', Toronto, Ontario, Canada'

for i in borough_names:
    geolocator = Nominatim(user_agent="toronto_explorer")
    location = geolocator.geocode(i + city_address)
    latitude = location.latitude
    longitude = location.longitude

    print('The geographical coordinates of {} are {}, {}.'.format(i + city_address, latitude, longitude))

The geographical coordinates of Downtown Toronto, Toronto, Ontario, Canada are 43.6563221, -79.3809161.
The geographical coordinates of East Toronto, Toronto, Ontario, Canada are 43.6247901, -79.3934918.
The geographical coordinates of West Toronto, Toronto, Ontario, Canada are 43.6534817, -79.3839347.
The geographical coordinates of Central Toronto, Toronto, Ontario, Canada are 43.6534817, -79.3839347.


<a id="folium_boroughs"></a>

Now we can visualize the neighborhoods for all boroughs with the name 'Toronto' in them.

In [30]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(borough_data['Latitude'], \
                                           borough_data['Longitude'], \
                                           borough_data['Borough'], \
                                           borough_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

<a id="foursquare_api"></a>

Next, we are going to utilize the Foursquare API to explore the neighborhoods and segment them.

First, we define our Foursquare credentials for calling the API. (Note: I have hidden my actual credentials, so if the notebook is run as-is, it will fail after this point.)

In [56]:
CLIENT_ID = 'GE0K3UT4PQX3EUVBJXLXSWHW1FV0DLRUJVFIGR2OO1WTLS1F' # your Foursquare ID
CLIENT_SECRET = 'R4K0XHQK4TZZ3GBFGDKAGF2JVHKMCX25D43DDBQB25NGSB2I' # your Foursquare Secret

In [34]:
CLIENT_ID = '...' # your Foursquare ID
CLIENT_SECRET = '...' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
print("Saved Foursquare credentials!")

Saved Foursquare credentials!


Next, we're going to grab the top 100 venues that are in each of the neighborhoods associated with our filtered borough names within a radius of 500 meters.

To do this, we are going to borrow the custom function, ```getNearbyVenues```, to get these venues.

In [35]:
LIMIT = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]["groups"][0]["items"]
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [36]:
# Get the associated venues information.
toronto_venues = getNearbyVenues(names=borough_data['Neighborhood'],
                                 latitudes=borough_data['Latitude'],
                                 longitudes=borough_data['Longitude']
                                )

Regent Park, Harbourfront
Queen's Park, Ontario Provincial Government
Garden District, Ryerson
St. James Town
The Beaches
Berczy Park
Central Bay Street
Christie
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
The Danforth West, Riverdale
Toronto Dominion Centre, Design Exchange
Brockton, Parkdale Village, Exhibition Place
India Bazaar, The Beaches West
Commerce Court, Victoria Hotel
Studio District
Lawrence Park
Roselawn
Davisville North
Forest Hill North & West
High Park, The Junction South
North Toronto West
The Annex, North Midtown, Yorkville
Parkdale, Roncesvalles
Davisville
University of Toronto, Harbord
Runnymede, Swansea
Moore Park, Summerhill East
Kensington Market, Chinatown, Grange Park
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Rosedale
Stn A PO Boxes
St. James Town,

In [37]:
print(toronto_venues.shape)
toronto_venues.head()

(1623, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  Regent Park, Harbourfront               43.65426              -79.360636   
1  Regent Park, Harbourfront               43.65426              -79.360636   
2  Regent Park, Harbourfront               43.65426              -79.360636   
3  Regent Park, Harbourfront               43.65426              -79.360636   
4  Regent Park, Harbourfront               43.65426              -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2      Morning Glory Cafe       43.653947       -79.361149   
3  Cooper Koo Family YMCA       43.653249       -79.358008   
4     Body Blitz Spa East       43.654735       -79.359874   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2       Breakfast Spot  
3  Distribution Center  
4                  Spa

We can see that a total of 1,623 venues were returned.

Now, let's group these neighborhoods, count them, and see how many unique categories of venues there are for our subset.

In [38]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            57   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing CentrE                                  18   
CN Tower, King and Spadina, Railway Lands, Harb...                     17   
Central Bay Street                                                     63   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             36   
Davisville North                                                        8   
Dufferin, Dovercourt Village                                           19   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              55   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               41   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     20   
Parkdale, Roncesvalles                                                 13   
Queen's Park, Ontario Provincial Government                            38   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               94   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede, Swansea                                                     39   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         95   
Studio District                                                        41   
Summerhill West, Rathnelly, South Hill, Forest ...                     17   
The Annex, North Midtown, Yorkville                                    22   
The Beaches                                                             4   
The Danforth West, Riverdale                                           43   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         36   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             57   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing CentrE                                   18   
CN Tower, King and Spadina, Railway Lands, Harb...                      17   
Central Bay Street                                                      63   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              36   
Davisville North                    

In [39]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 231 uniques categories.


<a id="neighborhood_analysis"></a>

Now we get to analyze each neighborhood.

We start by using one-hot encoding on venue categories.

In [40]:
# one-hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Yoga Studio  Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0            0        0                   0             0               0   
1            0        0                   0             0               0   
2            0        0                   0             0               0   
3            0        0                   0             0               0   
4            0        0                   0             0               0   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0                0                 0                    0             0   
1                0                 0                    0             0   
2                0                 0                    0             0   
3                0                 0                    0             0   
4                0                 0                    0             0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0         0            0           0                    0                 0   
1         0            0           0                    0                 0   
2         0            0           0                    0                 0   
3         0            0           0                    0                 0   
4         0            0           0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       1     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               1   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0              0         0   

   Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0        0     0                          0            0   
1        0     0                          0            0   
2        0     0                          0            0   
3        0     0                          0            0   
4        0     0                          0            0   

In [41]:
toronto_onehot.shape

(1623, 231)

Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category.

In [42]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio   Airport  \
0                                         Berczy Park     0.000000  0.000000   
1        Brockton, Parkdale Village, Exhibition Place     0.000000  0.000000   
2               Business reply mail Processing CentrE     0.055556  0.000000   
3   CN Tower, King and Spadina, Railway Lands, Har...     0.000000  0.058824   
4                                  Central Bay Street     0.015873  0.000000   
5                                            Christie     0.000000  0.000000   
6                                Church and Wellesley     0.025641  0.000000   
7                      Commerce Court, Victoria Hotel     0.000000  0.000000   
8                                          Davisville     0.000000  0.000000   
9                                    Davisville North     0.000000  0.000000   
10                       Dufferin, Dovercourt Village     0.000000  0.000000   
11             First Canadian Place, Underground city     0.000000  0.000000   
12                           Forest Hill North & West     0.000000  0.000000   
13                           Garden District, Ryerson     0.000000  0.000000   
14  Harbourfront East, Union Station, Toronto Islands     0.000000  0.000000   
15                      High Park, The Junction South     0.000000  0.000000   
16                     India Bazaar, The Beaches West     0.000000  0.000000   
17          Kensington Market, Chinatown, Grange Park     0.000000  0.000000   
18                                      Lawrence Park     0.000000  0.000000   
19                           Little Portugal, Trinity     0.024390  0.000000   
20                        Moore Park, Summerhill East     0.000000  0.000000   
21                                 North Toronto West     0.050000  0.000000   
22                             Parkdale, Roncesvalles     0.000000  0.000000   
23        Queen's Park, Ontario Provincial Government     0.026316  0.000000   
24                          Regent Park, Harbourfront     0.020833  0.000000   
25                           Richmond, Adelaide, King     0.000000  0.000000   
26                                           Rosedale     0.000000  0.000000   
27                                           Roselawn     0.000000  0.000000   
28                                 Runnymede, Swansea     0.025641  0.000000   
29                                     St. James Town     0.000000  0.000000   
30                        St. James Town, Cabbagetown     0.000000  0.000000   
31                                     Stn A PO Boxes     0.010526  0.000000   
32                                    Studio District     0.024390  0.000000   
33  Summerhill West, Rathnelly, South Hill, Forest...     0.000000  0.000000   
34                The Annex, North Midtown, Yorkville     0.000000  0.000000   
35                                        The Beaches     0.000000  0.000000   
36                       The Danforth West, Riverdale     0.023256  0.000000   
37           Toronto Dominion Centre, Design Exchange     0.000000  0.000000   
38                     University of Toronto, Harbord     0.027778  0.000000   

    Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0             0.000000      0.000000        0.000000         0.000000   
1             0.000000      0.000000        0.000000         0.000000   
2             0.000000      0.000000        0.000000         0.000000   
3             0.058824      0.058824        0.117647         0.176471   
4             0.000000      0.000000        0.000000         0.000000   
5             0.000000      0.000000        0.000000         0.000000   
6             0.000000      0.000000        0.000000         0.000000   
7             0.000000      0.000000        0.000000         0.000000   
8             0.000000      0.000000        0.000000         0.000000   
9             0.000000      0.000000        0.000000         0.000000   
10            0.000000      0.000000 

Let's check out the new size.

In [43]:
toronto_grouped.shape

(39, 231)

Let's print each neighborhood along with the top 5 most common venues.

In [44]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Berczy Park----
                venue  freq
0         Coffee Shop  0.07
1        Cocktail Bar  0.05
2                Café  0.04
3          Restaurant  0.04
4  Seafood Restaurant  0.04


----Brockton, Parkdale Village, Exhibition Place----
            venue  freq
0            Café  0.13
1  Breakfast Spot  0.09
2       Nightclub  0.09
3     Coffee Shop  0.09
4   Grocery Store  0.04


----Business reply mail Processing CentrE----
              venue  freq
0       Yoga Studio  0.06
1     Auto Workshop  0.06
2       Pizza Place  0.06
3        Comic Shop  0.06
4  Recording Studio  0.06


----CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport----
              venue  freq
0   Airport Service  0.18
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3          Boutique  0.06
4  Sculpture Garden  0.06


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.17
1  Italian Restaurant  0.08
2                Café  

<a id="top_venues_dataframe"></a>

Let's put that into a _pandas_ dataframe.

First, let's write a function to sort the venues in descending order.

In [45]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [46]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2              Business reply mail Processing CentrE           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar              Beer Bar           Cheese Shop   
1           Coffee Shop             Nightclub        Breakfast Spot   
2         Auto Workshop                  Park            Comic Shop   
3        Airport Lounge      Airport Terminal      Sculpture Garden   
4                  Café    Italian Restaurant        Sandwich Place   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                       Café                Bakery   
1             Pet Store                    Stadium          Intersection   
2           Pizza Place           Recording Studio         Burrito Place   
3              Boutique        Rental Car Location                 Plane   
4        Ice Cream Shop  Middle Eastern Restaurant                   Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0    Seafood Restaurant                Bistro              Jazz Club  
1                Bakery    Italian Restaurant             Restaurant  
2            Restaurant               Brewery     Light Rail Station  
3         Boat or Ferry       Harbor / Marina                    Bar  
4       Thai Restaurant          Burger Joint    Fried Chicken Joint

<a id="clustering_neighborhoods"></a>

Now we can work toward clustering our neighborhoods.

Run *k*-means to cluster the neighborhoods into 5 clusters.

In [47]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [48]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = borough_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode           Borough                                 Neighborhood  \
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
9         M5B  Downtown Toronto                     Garden District, Ryerson   
15        M5C  Downtown Toronto                               St. James Town   
19        M4E      East Toronto                                  The Beaches   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
2   43.654260 -79.360636               0           Coffee Shop   
4   43.662301 -79.389494               0           Coffee Shop   
9   43.657162 -79.378937               0        Clothing Store   
15  43.651494 -79.375418               0                  Café   
19  43.676357 -79.293031               0                 Trail   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
2                   Park                Bakery                   Pub   
4       Sushi Restaurant                 Diner           Yoga Studio   
9            Coffee Shop                  Café        Cosmetics Shop   
15           Coffee Shop          Cocktail Bar   American Restaurant   
19     Health Food Store                   Pub         Women's Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
2         Breakfast Spot                  Café                      Theater   
4          Burrito Place              Beer Bar           Italian Restaurant   
9             Restaurant   Japanese Restaurant           Italian Restaurant   
15             Gastropub                 Hotel                   Restaurant   
19          Dance Studio     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue      9th Most Common Venue 10th Most Common Venue  
2     Mexican Restaurant                 Restaurant         Chocolate Shop  
4              Juice Bar             Sandwich Place           Burger Joint  
9        Bubble Tea Shop  Middle Eastern Restaurant       Ramen Restaurant  
15                   Gym         Italian Restaurant       Department Store  
19            Donut Shop           Doner Restaurant                Dog Run

Finally, let's visualize the resulting clusters

In [50]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

<a id="examine_clusters"></a>

Now, we can examine each cluster.

**Cluster 1**

In [51]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto               0           Coffee Shop   
4    Downtown Toronto               0           Coffee Shop   
9    Downtown Toronto               0        Clothing Store   
15   Downtown Toronto               0                  Café   
19       East Toronto               0                 Trail   
20   Downtown Toronto               0           Coffee Shop   
24   Downtown Toronto               0           Coffee Shop   
25   Downtown Toronto               0         Grocery Store   
30   Downtown Toronto               0           Coffee Shop   
31       West Toronto               0                Bakery   
36   Downtown Toronto               0           Coffee Shop   
37       West Toronto               0                   Bar   
41       East Toronto               0      Greek Restaurant   
42   Downtown Toronto               0           Coffee Shop   
43       West Toronto               0                  Café   
47       East Toronto               0        Sandwich Place   
48   Downtown Toronto               0           Coffee Shop   
54       East Toronto               0                  Café   
67    Central Toronto               0                  Park   
69       West Toronto               0    Mexican Restaurant   
73    Central Toronto               0           Coffee Shop   
74    Central Toronto               0                  Café   
75       West Toronto               0             Gift Shop   
79    Central Toronto               0          Dessert Shop   
80   Downtown Toronto               0                  Café   
81       West Toronto               0           Pizza Place   
84   Downtown Toronto               0                  Café   
86    Central Toronto               0                   Pub   
87   Downtown Toronto               0       Airport Service   
92   Downtown Toronto               0           Coffee Shop   
96   Downtown Toronto               0           Coffee Shop   
97   Downtown Toronto               0           Coffee Shop   
99   Downtown Toronto               0           Coffee Shop   
100      East Toronto               0           Yoga Studio   

    2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
2                    Park                Bakery                          Pub   
4        Sushi Restaurant                 Diner                  Yoga Studio   
9             Coffee Shop                  Café               Cosmetics Shop   
15            Coffee Shop          Cocktail Bar          American Restaurant   
19      Health Food Store                   Pub                Women's Store   
20           Cocktail Bar              Beer Bar                  Cheese Shop   
24                   Café    Italian Restaurant               Sandwich Place   
25                   Café                  Park                  Candy Store   
30                   Café            Restaurant              Thai Restaurant   
31               Pharmacy           Supermarket                  Music Venue   
36               Aquarium                 Hotel                         Café   
37             Restaurant           Men's Store        Vietnamese Restaurant   
41            Coffee Shop    Italian Restaurant               Ice Cream Shop   
42                  Hotel                  Café                   Restaurant   
43            Coffee Shop             Nightclub               Breakfast Spot   
47   Fast Food Restaurant         Movie Theater                          Pub   
48                   Café            Restaurant                        Hotel   
54            Coffee Shop             Gastropub                       Bakery   
67       Department Store                   Gym               Breakfast Spot   
69                   Café       Thai Restaurant         Fast Food Restaurant   
73         Clothing Store                  Park           Salon / Barbershop   
74         Sandwich Place           Coffee Shop 

**Cluster 2**

In [52]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto               1                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62         Women's Store         Deli / Bodega           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
62  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62            Donut Shop      Doner Restaurant                Dog Run

**Cluster 3**

In [53]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
83  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
83         Women's Store         Deli / Bodega  Ethiopian Restaurant   

   5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
83     Electronics Store  Eastern European Restaurant            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
83      Doner Restaurant               Dog Run    Distribution Center

**Cluster 4**

In [54]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
68   Central Toronto               3                  Park   
91  Downtown Toronto               3                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
68         Jewelry Store                 Trail              Bus Line   
91            Playground                 Trail         Women's Store   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
68      Sushi Restaurant         Deli / Bodega            Electronics Store   
91          Dance Studio     Electronics Store  Eastern European Restaurant   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
68  Eastern European Restaurant            Donut Shop       Doner Restaurant  
91                   Donut Shop      Doner Restaurant                Dog Run

**Cluster 5**

In [55]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
61  Central Toronto               4           Swim School   

   2nd Most Common Venue       3rd Most Common Venue 4th Most Common Venue  \
61                  Park  Construction & Landscaping              Bus Line   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
61         Women's Store  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
61  Eastern European Restaurant            Donut Shop       Doner Restaurant